File: tools/state_text.py
"""Creature state -> the user-turn text Needle reads. Mirrors
firmware/app/state_text.cpp byte for byte (tests/test_state_text.py checks).

The numbers are the app's state; turning them into a sentence is the app's
job, the same way the Doom project (058) renders its game state as text.
Needle then decides from that text alone. No rule here picks a tool.
"""

HUNGER = [(20, "not hungry"), (50, "a little hungry"), (75, "hungry"), (100, "starving")]
ENERGY = [(20, "exhausted"), (40, "tired"), (70, "rested"), (100, "full of energy")]
CURIOSITY = [(30, "bored"), (60, "a bit curious"), (100, "very curious")]


def word(table, v):
    for top, w in table:
        if v <= top:
            return w
    return table[-1][1]


def render(state, template="words"):
    h, e, c = state["hunger"], state["energy"], state["curiosity"]
    if template == "numbers":
        return f"hunger: {h}\nenergy: {e}\ncuriosity: {c}"
    if template == "scores":
        return f"Hunger {h}/100, energy {e}/100, curiosity {c}/100. What should I do?"
    if template == "words":
        return f"I'm {word(HUNGER, h)}, {word(ENERGY, e)} and {word(CURIOSITY, c)}."
    raise ValueError(template)


TEMPLATES = ("numbers", "scores", "words")

File: firmware/app/state_text.cpp
#include "state_text.h"

#include <stdio.h>

// Same buckets as tools/state_text.py (HUNGER, ENERGY, CURIOSITY).
const char* hunger_word(int v) {
    return v <= 20 ? "not hungry" : v <= 50 ? "a little hungry" : v <= 75 ? "hungry" : "starving";
}
const char* energy_word(int v) {
    return v <= 20 ? "exhausted" : v <= 40 ? "tired" : v <= 70 ? "rested" : "full of energy";
}
const char* curiosity_word(int v) {
    return v <= 30 ? "bored" : v <= 60 ? "a bit curious" : "very curious";
}

int state_text(const CreatureState& s, char* out, int cap) {
    return snprintf(out, cap, "I'm %s, %s and %s.", hunger_word(s.hunger), energy_word(s.energy),
                    curiosity_word(s.curiosity));
}
